File: sdk/csharp/RinoxAuth.cs
using System;
using System.Collections.Generic;
using System.Management;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace YourAuth
{
    // ============================================
    // AUTH CLIENT CLASS
    // ============================================
    public class AuthClient
    {
        // Properties
        public bool IsSuccess { get; private set; }
        public string Message { get; private set; }
        public UserInfo User { get; private set; }
        public string Token { get; private set; }

        // Private fields
        private readonly string _apiUrl;
        private readonly string _appName;
        private readonly string _ownerIdOrAppKey;
        private readonly string _secret;
        private readonly string _version;

        private static readonly HttpClient Http = new HttpClient();

        // ============================================
        // CONSTRUCTOR
        // ============================================
        public AuthClient(string appName, string ownerIdOrAppKey, string secret, string version, string apiBaseUrl)
        {
            _appName = appName;
            _ownerIdOrAppKey = ownerIdOrAppKey;
            _secret = secret;
            _version = version;
            _apiUrl = apiBaseUrl.TrimEnd('/') + "/api/auth";
        }

        // ============================================
        // INIT - Check connection
        // ============================================
        public async Task<bool> Init()
        {
            var payload = new Dictionary<string, object>
            {
                ["type"] = "init",
                ["appname"] = _appName,
                ["ownerid"] = _ownerIdOrAppKey,
                ["secret"] = _secret,
                ["version"] = _version
            };

            await SendAsync(payload);
            return IsSuccess;
        }

        // ============================================
        // LOGIN - Username & Password
        // ============================================
        public async Task<bool> Login(string username, string password)
        {
            var payload = new Dictionary<string, object>
            {
                ["type"] = "login",
                ["appname"] = _appName,
                ["ownerid"] = _ownerIdOrAppKey,
                ["secret"] = _secret,
                ["username"] = username,
                ["password"] = password,
                ["hwid"] = GetHwid()
            };

            await SendAsync(payload);
            return IsSuccess;
        }

        // ============================================
        // LICENSE - Validate license key
        // ============================================
        public async Task<bool> License(string key)
        {
            var payload = new Dictionary<string, object>
            {
                ["type"] = "license",
                ["appname"] = _appName,
                ["ownerid"] = _ownerIdOrAppKey,
                ["secret"] = _secret,
                ["key"] = key,
                ["hwid"] = GetHwid()
            };

            await SendAsync(payload);
            return IsSuccess;
        }

        // ============================================
        // VALIDATE LICENSE (Alias for License)
        // ============================================
        public async Task<bool> ValidateLicense(string key)
        {
            return await License(key);
        }

        // ============================================
        // ✅ LICENSE LOGIN - Login with license key only
        // ============================================
        public async Task<bool> LicenseLogin(string key)
        {
            var payload = new Dictionary<string, object>
            {
                ["type"] = "license-login",
                ["appname"] = _appName,
                ["ownerid"] = _ownerIdOrAppKey,
                ["secret"] = _secret,
                ["key"] = key,
                ["hwid"] = GetHwid()
            };

            await SendAsync(payload);
            return IsSuccess;
        }

        // ============================================
        // ✅ REGISTER
        // ============================================
        public async Task<bool> Register(string username, string password, string licenseKey = null)
        {
            var payload = new Dictionary<string, object>
            {
                ["type"] = "register",
                ["appname"] = _appName,
                ["ownerid"] = _ownerIdOrAppKey,
                ["secret"] = _secret,
                ["username"] = username,
                ["password"] = password,
                ["hwid"] = GetHwid()
            };

            if (!string.IsNullOrEmpty(licenseKey))
            {
                payload["key"] = licenseKey;
            }

            await SendAsync(payload);
            return IsSuccess;
        }

        // ============================================
        // ✅ QUICK LICENSE LOGIN (Simple one-call)
        // ============================================
        public async Task<bool> QuickLicenseLogin(string key)
        {
            // First try license-login
            if (await LicenseLogin(key))
            {
                return true;
            }

            // If fails, try validate + login
            if (await License(key))
            {
                string username = "user_" + key.Replace("-", "").Substring(0, Math.Min(8, key.Replace("-", "").Length));
                return await Login(username, key);
            }

            return false;
        }

        // ============================================
        // CORE SEND METHOD
        // ============================================
        private async Task SendAsync(Dictionary<string, object> payload)
        {
            try
            {
                var json = JsonConvert.SerializeObject(payload);
                var body = new StringContent(json, Encoding.UTF8, "application/json");

                var response = await Http.PostAsync(_apiUrl, body);
                var result = await response.Content.ReadAsStringAsync();

                var api = JsonConvert.DeserializeObject<ApiResponse>(result);

                IsSuccess = api.success;
                Message = api.message;
                User = api.user;
                Token = api.token;
            }
            catch (Exception ex)
            {
                IsSuccess = false;
                Message = ex.Message;
                User = null;
                Token = null;
            }
        }

        // ============================================
        // GET HWID
        // ============================================
        public static string GetHwid()
        {
            try
            {
                // Try CPU ID
                using (var searcher = new ManagementObjectSearcher("SELECT ProcessorId FROM Win32_Processor"))
                {
                    foreach (ManagementObject obj in searcher.Get())
                    {
                        string cpuId = obj["ProcessorId"]?.ToString();
                        if (!string.IsNullOrEmpty(cpuId) && cpuId != "0000000000000000")
                        {
                            return cpuId;
                        }
                    }
                }

                // Fallback: Volume Serial
                using (var searcher = new ManagementObjectSearcher("SELECT VolumeSerialNumber FROM Win32_LogicalDisk WHERE DeviceID='C:'"))
                {
                    foreach (ManagementObject obj in searcher.Get())
                    {
                        return obj["VolumeSerialNumber"]?.ToString() ?? "UNKNOWN";
                    }
                }
            }
            catch { }

            // Ultimate fallback
            return Environment.MachineName + "_" + Environment.UserName;
        }

        // ============================================
        // API RESPONSE CLASS
        // ============================================
        private class ApiResponse
        {
            public bool success { get; set; }
            public string message { get; set; }
            public UserInfo user { get; set; }
            public string token { get; set; }
        }
    }

    // ============================================
    // USER INFO CLASS
    // ============================================
    public class UserInfo
    {
        public string username { get; set; }
        public string expiry { get; set; }
        public string hwid { get; set; }
        public string plan { get; set; }
        public string role { get; set; }

        public bool IsExpired()
        {
            if (string.IsNullOrEmpty(expiry) || expiry == "never" || expiry == "Never") return false;
            if (DateTime.TryParse(expiry, out DateTime expiryDate))
            {
                return expiryDate < DateTime.Now;
            }
            return false;
        }

        public int DaysRemaining()
        {
            if (string.IsNullOrEmpty(expiry) || expiry == "never" || expiry == "Never") return 999;
            if (DateTime.TryParse(expiry, out DateTime expiryDate))
            {
                return (expiryDate - DateTime.Now).Days;
            }
            return 0;
        }
    }
}

File: sdk/python/RinoxAuth.py
"""
RinoxAuth Python SDK
Modern licensing + authentication platform
"""

import json
import platform
import subprocess
import uuid
from typing import Optional, Dict, Any

try:
    import requests
except ImportError:
    print("Please install requests: pip install requests")
    raise


class UserInfo:
    """User information returned by the API"""
    
    def __init__(self, data: Dict[str, Any]):
        self.username: str = data.get("username", "")
        self.expiry: str = data.get("expiry", "")
        self.hwid: str = data.get("hwid", "")
        self.plan: str = data.get("plan", "free")
        self.role: str = data.get("role", "user")
    
    def is_expired(self) -> bool:
        """Check if user subscription has expired"""
        if not self.expiry or self.expiry in ("never", "Never", "N/A"):
            return False
        try:
            from datetime import datetime
            expiry_date = datetime.strptime(self.expiry[:10], "%Y-%m-%d")
            return expiry_date < datetime.now()
        except:
            return False
    
    def days_remaining(self) -> int:
        """Get days remaining until expiry"""
        if not self.expiry or self.expiry in ("never", "Never", "N/A"):
            return 999
        try:
            from datetime import datetime
            expiry_date = datetime.strptime(self.expiry[:10], "%Y-%m-%d")
            return (expiry_date - datetime.now()).days
        except:
            return 0
    
    def __repr__(self):
        return f"UserInfo(username={self.username}, plan={self.plan}, expiry={self.expiry})"


class AuthClient:
    """
    RinoxAuth Client for Python applications
    
    Usage:
        client = AuthClient(
            appName="atik",
            ownerIdOrAppKey="04fad000",
            secret="your-app-secret",
            version="1.0.0",
            apiBaseUrl="http://127.0.0.1:8000"
        )
        
        if client.init():
            if client.login("username", "password"):
                print("Login success!")
    """
    
    def __init__(
        self,
        appName: str,
        ownerIdOrAppKey: str,
        secret: str,
        version: str = "1.0.0",
        apiBaseUrl: str = "http://127.0.0.1:8000"
    ):
        """
        Initialize AuthClient
        
        Args:
            appName: Application name (from dashboard)
            ownerIdOrAppKey: Owner ID (from credentials)
            secret: App Secret (from credentials)
            version: App version (e.g., "1.0.0")
            apiBaseUrl: Backend URL (e.g., "http://127.0.0.1:8000")
        """
        self._app_name = appName
        self._owner_id = ownerIdOrAppKey
        self._secret = secret
        self._version = version
        self._api_url = apiBaseUrl.rstrip('/') + "/api/auth"
        
        # Results
        self.is_success: bool = False
        self.message: str = ""
        self.user: Optional[UserInfo] = None
        self.token: Optional[str] = None
    
    # ============================================
    # INIT
    # ============================================
    def init(self) -> bool:
        """Initialize connection and check credentials"""
        payload = {
            "type": "init",
            "appname": self._app_name,
            "ownerid": self._owner_id,
            "secret": self._secret,
            "version": self._version
        }
        self._send(payload)
        return self.is_success
    
    # ============================================
    # LOGIN (Username + Password)
    # ============================================
    def login(self, username: str, password: str) -> bool:
        """Login with username and password"""
        payload = {
            "type": "login",
            "appname": self._app_name,
            "ownerid": self._owner_id,
            "secret": self._secret,
            "username": username,
            "password": password,
            "hwid": self._get_hwid()
        }
        self._send(payload)
        return self.is_success
    
    # ============================================
    # LICENSE VALIDATION
    # ============================================
    def license(self, key: str) -> bool:
        """Validate a license key"""
        payload = {
            "type": "license",
            "appname": self._app_name,
            "ownerid": self._owner_id,
            "secret": self._secret,
            "key": key,
            "hwid": self._get_hwid()
        }
        self._send(payload)
        return self.is_success
    
    def validate_license(self, key: str) -> bool:
        """Alias for license()"""
        return self.license(key)
    
    # ============================================
    # LICENSE LOGIN (No username/password needed)
    # ============================================
    def license_login(self, key: str) -> bool:
        """Login using only a license key"""
        payload = {
            "type": "license-login",
            "appname": self._app_name,
            "ownerid": self._owner_id,
            "secret": self._secret,
            "key": key,
            "hwid": self._get_hwid()
        }
        self._send(payload)
        return self.is_success
    
    # ============================================
    # QUICK LICENSE LOGIN
    # ============================================
    def quick_license_login(self, key: str) -> bool:
        """
        One-call license login
        Tries license-login first, then falls back to validate + login
        """
        # First try license-login
        if self.license_login(key):
            return True
        
        # If fails, validate then login
        if self.license(key):
            username = "user_" + key.replace("-", "")[:8]
            return self.login(username, key)
        
        return False
    
    # ============================================
    # REGISTER
    # ============================================
    def register(self, username: str, password: str, license_key: Optional[str] = None) -> bool:
        """Register a new user account"""
        payload = {
            "type": "register",
            "appname": self._app_name,
            "ownerid": self._owner_id,
            "secret": self._secret,
            "username": username,
            "password": password,
            "hwid": self._get_hwid()
        }
        
        if license_key:
            payload["key"] = license_key
        
        self._send(payload)
        return self.is_success
    
    # ============================================
    # CORE SEND METHOD
    # ============================================
    def _send(self, payload: Dict[str, Any]) -> None:
        """Send request to API"""
        try:
            response = requests.post(
                self._api_url,
                json=payload,
                headers={"Content-Type": "application/json"},
                timeout=30
            )
            
            data = response.json()
            
            self.is_success = data.get("success", False)
            self.message = data.get("message", "")
            self.token = data.get("token")
            
            # Parse user info
            user_data = data.get("user") or data.get("data", {}).get("user")
            if user_data:
                self.user = UserInfo(user_data)
            else:
                self.user = None
                
        except requests.exceptions.ConnectionError:
            self.is_success = False
            self.message = "Connection failed. Is the server running?"
            self.user = None
            self.token = None
        except requests.exceptions.Timeout:
            self.is_success = False
            self.message = "Request timed out"
            self.user = None
            self.token = None
        except Exception as e:
            self.is_success = False
            self.message = str(e)
            self.user = None
            self.token = None
    
    # ============================================
    # GET HWID
    # ============================================
    @staticmethod
    def _get_hwid() -> str:
        """Get unique hardware ID"""
        try:
            # Try CPU serial
            if platform.system() == "Windows":
                result = subprocess.run(
                    ["wmic", "cpu", "get", "ProcessorId"],
                    capture_output=True, text=True
                )
                lines = result.stdout.strip().split('\n')
                if len(lines) > 1:
                    cpu_id = lines[1].strip()
                    if cpu_id and cpu_id != "0000000000000000":
                        return cpu_id
            else:
                # Linux/Mac
                result = subprocess.run(
                    ["cat", "/proc/cpuinfo"],
                    capture_output=True, text=True
                )
                for line in result.stdout.split('\n'):
                    if "Serial" in line:
                        return line.split(":")[1].strip()
        except:
            pass
        
        # Fallback: Machine UUID
        try:
            return str(uuid.getnode())
        except:
            pass
        
        # Ultimate fallback
        return f"{platform.node()}_{platform.machine()}"